### PR TITLE
Fix broken assertion in ContractCompiler

### LINKED ISSUE #8406: [solc] ContractCompiler throws ICE because "*type == *variable->annotation().type" fails
## Description

```
contract C {
  address constant e = 0x1212121212121212121212121000002134593163;

  function f() public returns (byte z) {
    assembly { z := e }
  }
}
```

## Environment

- Compiler version: latest develop
- Target EVM version (as per compiler settings): NA

## Steps to Reproduce

```
$ solc <input.sol>
Internal compiler error during compilation:
/home/bhargava/work/github/solidity/libsolidity/codegen/ContractCompiler.cpp(703): Throw in function auto solidity::frontend::ContractCompiler::visit(const solidity::frontend::InlineAssembly &)::(anonymous class)::operator()(const yul::Identifier &, yul::IdentifierContext, yul::AbstractAssembly &) const
Dynamic exception type: boost::exception_detail::clone_impl<solidity::langutil::InternalCompilerError>
std::exception::what: 
[solidity::util::tag_comment*] = 
```

### PR DESCRIPTION
**Pull request by randomnetcat**
Fixes #8406.

### CODE REVIEW

File: test/libsolidity/syntaxTests/constants/regression_8406.sol [hunk 1]
```diff
@@ -0,0 +1,12 @@
+// Test for regression of https://github.com/ethereum/solidity/issues/8406
```
Comment (chriseth) [inline test/libsolidity/syntaxTests/constants/regression_8406.sol]: Can you turn this into a semantic test?

**Comment by chriseth** [comment]
Please squash

**Comment by randomnetcat** [comment]
Done.

**Review by chriseth — DISMISSED** [review]

File: test/libsolidity/semanticTests/constants/regression_8406.sol [hunk 2]
```diff
@@ -0,0 +1,12 @@
+// Test for regression of https://github.com/ethereum/solidity/issues/8406
```
Comment (chriseth) [inline test/libsolidity/semanticTests/constants/regression_8406.sol]: Can you change the name of the test to be more descriptive?

Comment (randomnetcat) [reply test/libsolidity/semanticTests/constants/regression_8406.sol]: Renamed to `asm_address_constant_regression`.

**Review by chriseth — APPROVED** [review]

**Pull request merged**